### DISCUSSION TITLE
Bug in Chess App

### DISCUSSION THREAD

**Discussion by marte1337** [neuefische/questions#594] [Web] [answered]
## 🎉 Introduction:

I´m trying to integrate a "Zombie-Mode" into my chess app: if a white piece captures a black piece, the captured piece respawns as a white piece and vice versa. The respawn-square is the square, where the attacking piece was placed before the capture (f.e. if a white knight moved from f3 to e5 and captured a black bishop there, then a white bishop will respawn on f3).

This is how it works:

![Screenshot 2023-05-09 at 14 48 24](https://github.com/neuefische/questions/assets/54943747/cd8b0626-6752-4311-9833-08cba231cb95)

previousMoves is a state, that gathers all information about the move that just happened, including information about the squares, from where the referred piece was coming from and to where it went, and also if a capture took place (flags: "c") information about the captured piece (f.e. captured: "b").

In the code above, I use the previousMove object to check, if a piece was captured, and if true:

1. create a new game object, 
2. insert a fen(), which is basically a screenshot of the game. See https://www.chess.com/terms/fen-chess
3. add the zombie-piece to the new game object (with the .put()-function)
4. update the game object state by setGame(newGame)

When black captures a white piece, everything works fine. When I capture with white, the respawn does happen but gets overwritten as soon as the following black move occurs. 

When I console.log the current fen + previousMove (already implemented in pull-request), the problem gets visible: when a white piece captures black, the zombie piece doesn´t display in the fen, but when a black piece captures white, the zombie piece gets displayed and is therefore part of the game.

Also something that I don´t understand yet: When a capture takes place two console.logs (and therefore fens) are triggered, instead of one.  EDIT: This has to do with the double trigger Problem stated in the Comments below!

I tried to move the if-statement to every other possible place in the code, didn´t help. The useEffect was necessary because outside of another function, the if statement triggered endless rerender.

## 🧮 Pull Request or link to code:

EDIT (updated): https://github.com/marte1337/capstone-project/pull/17

Everything takes place in the PlayerVsRandomMove.js

## ♻️ What are the steps to reproduce this issue?

Capture a black piece with a white piece.

## 🤔 What happens?

The respawn is happening but gets overwritten as soon as the following black move occurs. 

## 👉 What did you expect to happen?

The respawned white piece should stay on the board. 

## 🤖 Any logs, error output, etc?

I sometimes get an undefined for the props, but this seems to be another issue.

**Comment by NikRoe** [comment]
still investigating 👀

**Comment by marte1337** [comment]
I got it to work on a stale board (when I make moves for white and black manually), but still no luck implementing the random-move-engine. But I think I found the bug at least: 

![Screenshot 2023-05-10 at 15 23 32](https://github.com/neuefische/questions/assets/54943747/65b5453b-a06f-45d4-abd5-1efcd6ec28f9)

` setTimeout(makeRandomMove, 800)` needs to get separated from onDrop and only triggered after a manual move is finished.

But I don´t know how to fix that issue yet...

**Reply by NikRoe** [reply]
I'll take another look 👀

**Comment by marte1337** [comment]
Please see this new pull request dedicated specifically to the problem:

https://github.com/marte1337/capstone-project/pull/17

Everything takes place in components/Boards/PlayerVsRandomMove.js

**Comment by marte1337** [comment, marked as answer]
Fixed the issue! This is the solution:

<img width="776" alt="Screenshot 2023-05-14 at 20 26 31" src="https://github.com/neuefische/questions/assets/54943747/74a5f178-3ebc-42fc-8a35-d8d627de555f">

**Reply by NikRoe** [reply]
nice 👍
